### PR TITLE
ENGA-1457b: Pull Date Options

### PR DESCRIPTION
**Pull request by DevDaveFrame**
[ENGA-1457](https:/industrydive.atlassian.net/browse/ENGA-1457)

### Non-technical Context
This change adds timezone support to the BigQuery Exporter and makes the pull date field configurable, allowing for better control over exported data and more reliable timezone handling.

### Technical Context
1. **bigquery_exporter/base.py:** Added proper timezone handling to ensure all datetime objects are converted to UTC before export. Added two new class attributes (`include_pull_date` and `pull_date_field_name`) to control whether pull dates are included in exported data and customize the field name. Modified `_process_queryset()`, `_sanitize_value()`, and `table_has_data()` methods to support these features.

2. **requirements.txt and setup.py:** Added `pytz` as a dependency with version requirement `>= 2023.3` to support timezone operations.

3. **tests/test_base.py:** Added tests for:
   - Processing querysets with and without pull date inclusion
   - Custom pull date field naming
   - Naive and timezone-aware datetime handling
   - UTC conversion in the table_has_data method

### Manual Testing Instructions
1. Create a BigQueryExporter subclass with `include_pull_date = True` and verify pull dates are included in exports
2. Test with timezone-aware datetimes to confirm proper UTC conversion
3. Set a custom `pull_date_field_name` and verify the field appears with the specified name in the export

#### Developer

- [x] Ensure code complies with the [style guide](mdc:https:/industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
- [x] Devise edge cases to test the bug fix or feature being merged
- [x] Make sure the code is clean and understandable
- [x] Ensure code is not overly inefficient
- [ ] Ensure user permissions have been included (for new features) or updated (for revisions), if necessary
- [ ] Ensure documentation is understandable and accurate, including:
    - [ ] Code comments and doc strings
    - [ ] Technical documentation

#### Reviewer
- [x] Ensure code complies with the [style guide](mdc:https:/industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
    - If there are few or minor issues, add inline comments
    - If there are many issues or it seems like the developer did not follow the style guide, leave a comment asking them to do so
- [x] Devise edge cases to test the bug fix or feature being merged
- [x] Make sure the code is clean and understandable
    - [ ] Ask for inline comments on unclear sections
- [x] Ensure code is not overly inefficient
- [x] Ensure user permissions have been included (for new features) or updated (for revisions), if necessary
- [x] Ensure documentation is understandable and accurate, including:
    - [x] Code comments and doc strings
    - [ ] Technical documentation
- [ ] There are no conflicting migration leaf nodes

[ENGA-1457]: https://industrydive.atlassian.net/browse/ENGA-1457?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bigquery_exporter/base.py [hunk 1]
```diff
@@ -168,8 +171,15 @@ def table_has_data(self, pull_date=None):
...
+            # Convert pull_date to UTC if it has timezone info
+            if pull_date.tzinfo is not None:
+                utc_date = pull_date.astimezone(pytz.UTC)
+            else:
+                # If naive datetime (no timezone), assume it's in UTC
+                utc_date = pytz.UTC.localize(pull_date)
+
+            query = f'SELECT COUNT(*) FROM {self.table_name} WHERE DATE({self.pull_date_field_name}) = "{utc_date.strftime("%Y-%m-%d")}"'
         else:
             query = f'SELECT COUNT(*) FROM {self.table_name}'
```
Comment (dpuletic) [inline bigquery_exporter/base.py:186]: @DevDaveFrame , why is this query checking if there is any data in the table regardless of date it was added? Shouldn't it include only data with pull_date=today in this case?

File: bigquery_exporter/base.py [hunk 2]
```diff
@@ -201,7 +211,9 @@ def _push_to_bigquery(self, data, retry_deadline=600):
     def _process_queryset(self, queryset, pull_time):
         processed_queryset = []
         for model_instance in queryset:
-            processed_dict = {'pull_date': pull_time.strftime('%Y-%m-%d %H:%M:%S')}
+            processed_dict = {}
+            if self.include_pull_date:
+                processed_dict[self.pull_date_field_name] = self._sanitize_value(pull_time)
```
Comment (dpuletic) [inline bigquery_exporter/base.py]: How about we rename the `pull_time` variable to `pull_datetime` or something similar to better match the variable type? We could apply the same renaming in the `export` method where this helper is called.

File: bigquery_exporter/base.py [hunk 3]
```diff
@@ -227,7 +240,13 @@ def _sanitize_value(self, value):
...
         """
         if isinstance(value, datetime.datetime):
-            return value.strftime('%Y-%m-%d %H:%M:%S')
+            # Convert datetime to UTC if it has timezone info
+            if value.tzinfo is not None:
+                utc_value = value.astimezone(pytz.UTC)
+            else:
+                # If naive datetime (no timezone), assume it's in UTC
+                utc_value = pytz.UTC.localize(value)
+            return utc_value.strftime('%Y-%m-%d %H:%M:%S')
```
Comment (dpuletic) [inline bigquery_exporter/base.py]: How about we extract the code block that you added into a little helper function? Something like:
```
    def handle_datetime_value(value): # feel free to rename it to whatever you think makes more sense
        # Convert datetime to UTC if it has timezone info
        if value.tzinfo is not None:
            utc_value = value.astimezone(pytz.UTC)
        else:
            # If naive datetime (no timezone), assume it's in UTC
            utc_value = pytz.UTC.localize(value)
        return utc_value.strftime('%Y-%m-%d %H:%M:%S')
```
That way, the `_sanitize_value` function would look super clean :)
```
    def _sanitize_value(self, value):
        if isinstance(value, datetime.datetime):
            return handle_datetime_value(value)
        elif isinstance(value, UUID):
            return str(value)
        elif value is None:
            return '' if self.replace_nulls_with_empty else None
        else:
            return value
```

**Review by dpuletic — COMMENTED** [review]
@DevDaveFrame , this looks great! I just added couple of refactoring suggestions and had one question for my own understanding.

File: bigquery_exporter/base.py (hunk 1 above)
Comment (DevDaveFrame) [reply bigquery_exporter/base.py:186]: Yeah my thought was that if we don't specify a date for which we want to check data, we just check if the table has any data at all, which might be useful for some use cases. It also handles the new case where include_pull_date is false in which case there is presumed to be no field to compare a date against. I remember the former as the original motivation, but the docstring does state otherwise, so not sure when the mismatch occurred. If you think it makes sense, I can just update the docstring

File: bigquery_exporter/base.py (hunk 2 above)
Comment (DevDaveFrame) [reply bigquery_exporter/base.py]: Yeah that makes sense will do!

File: bigquery_exporter/base.py (hunk 3 above)
Comment (DevDaveFrame) [reply bigquery_exporter/base.py]: Ah yeah absolutely!

File: bigquery_exporter/base.py (hunk 1 above)
Comment (dpuletic) [reply bigquery_exporter/base.py:186]: Yup. That makes sense! Thanks for explaining.

**Comment by dpuletic** [comment]
@DevDaveFrame , thanks for updating the code based on my suggestions!

**Review by dpuletic — APPROVED** [review]
